### PR TITLE
Rewriting channel based on new gate component 

### PR DESCRIPTION
**Pull request by hkaiser**
The new `gate` component is a derivative of `and_gate`. Adding test.

### CODE REVIEW

File: hpx/lcos/local/gate.hpp [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+                return promise_.get_future(ec);
+            }
+            return hpx::future<void>();
+        }
+
+        /// \brief Set the data which has to go into the segment \a which.
+        template <typename OuterLock>
+        void set(OuterLock & outer_lock, error_code& ec = throws)
+        {
+            boost::unique_lock<mutex_type> l(mtx_);
```
Comment (sithhell) [inline hpx/lcos/local/gate.hpp]: This will potentially lead to held lock during suspension.

File: tests/unit/lcos/channel.cpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        hpx::lcos::local::channel<int> c;
+
+        hpx::future<void> g = hpx::async(&receive_values, boost::ref(c));
+        send_values(c);
+
+        g.get();
+    }
+
+    // two calls to get_future() without waiting for the first one to become
+    // ready should throw
```
Comment (sithhell) [inline tests/unit/lcos/channel.cpp]: I'm not sure if that is something that's practical. Consider something like that:
```
while(...)
{
    c.get_future().then(...);
}
```

With that restriction, I think code like that is not safe.

**Comment by hkaiser** [comment]
Looks like the design of this is broken, might be better to remove `channel` alltogether. Even more as we already have the `receive_buffer`. Thoughts?

**Comment by hkaiser** [comment]
I'm going to close this without merging as I don't know how to fix it properly at this point.

**Pull request closed without merging**